pinot-plugins/pinot-file-system/pinot-gcs/src/main/java/org/apache/pinot/plugin/filesystem/GcsPinotFS.java

```diff
@@ -173,228 +310,138 @@ private boolean isEmptyDirectory(URI uri) throws IOException {
-      requireNonNull(uri, "uri is null");
-      String path = normalizeToDirectoryPrefix(uri);
-      // Bucket root directory already exists and cannot be created
-      if (path.equals(DELIMITER)) {
-        return true;
+      ImmutableList.Builder<String> builder = ImmutableList.builder();
+      String prefix = fileUri.getPrefix();
+      Page<Blob> page;
+      if (recursive) {
+        page = _storage.list(fileUri.getBucketName(), Storage.BlobListOption.prefix(prefix));
+      } else {
+        page = _storage.list(fileUri.getBucketName(), Storage.BlobListOption.prefix(prefix),
+            Storage.BlobListOption.currentDirectory());
       }
-      Blob blob = getBucket(uri).create(normalizeToDirectoryPrefix(uri), new byte[0]);
-      return blob.exists();
-    } catch (Throwable t) {
+      page.iterateAll().forEach(blob -> {
+        if (!blob.getName().equals(prefix)) {
+          builder.add(GcsUri.createGcsUri(fileUri.getBucketName(), blob.getName()).toString());
+        }
+      });
+      String[] listedFiles = builder.build().toArray(new String[0]);
+      LOGGER.info("Listed {} files from URI: {}, is recursive: {}", listedFiles.length, fileUri, recursive);
+      return listedFiles;
+    } catch (Exception t) {
       throw new IOException(t);
     }
   }
 
-  @Override
-  public boolean delete(URI segmentUri, boolean forceDelete) throws IOException {
-    LOGGER.info("Deleting uri {} force {}", segmentUri, forceDelete);
+  private boolean exists(GcsUri gcsUri)
+      throws IOException {
+    if (existsDirectory(gcsUri)) {
+      return true;
+    }
+    if (isPathTerminatedByDelimiter(gcsUri)) {
+      return false;
+    }
+    return existsFile(gcsUri);
+  }
+
+  private boolean delete(GcsUri segmentUri, boolean forceDelete)
+      throws IOException {
     try {
       if (!exists(segmentUri)) {
-        return false;
+        return forceDelete;
       }
-      if (isDirectory(segmentUri)) {
-        if (!forceDelete) {
-          checkState(isEmptyDirectory(segmentUri), "ForceDelete flag is not set and directory '%s' is not empty", segmentUri);
+      if (existsDirectory(segmentUri)) {
+        if (!forceDelete && !isEmptyDirectory(segmentUri)) {
+          return false;
         }
-        String prefix = normalizeToDirectoryPrefix(segmentUri);
+        String prefix = segmentUri.getPrefix();
         Page<Blob> page;
-        if (prefix.equals(DELIMITER)) {
+        if (prefix.equals(GcsUri.DELIMITER)) {
           page = getBucket(segmentUri).list();
         } else {
           page = getBucket(segmentUri).list(Storage.BlobListOption.prefix(prefix));
         }
-        boolean deleteSucceeded = true;
-        for (Blob blob : page.iterateAll()) {
-          deleteSucceeded &= blob.delete();
-        }
-        return deleteSucceeded;
+        return batchDelete(page);
       } else {
         Blob blob = getBlob(segmentUri);
         return blob != null && blob.delete();
       }
     } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)